Update TestNoRestraint and TestWindup

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug

/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:
This PR tightens up some queueset tests, both because that is good on its own and in preparation for further work on queueset and its tests.

Make TestNoRestraint verify that fairness is NOT achieved
when there is real competition.

Make TestWindup run two cases, to show that 0.1 is too narrow
a margin and 0.26 is wide enough.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Update TestNoRestraint and TestWindup

Make TestNoRestraint verify that fairness is NOT achieved
when there is real competition.

Make TestWindup run two cases, to show that 0.1 is too narrow
a margin and 0.26 is wide enough.

MikeSpreitzer commented:
@kubernetes/sig-api-machinery-misc 
/cc @wojtek-t 
/cc @tkashem 
@yue9944882

MikeSpreitzer commented:
@deads2k 
@lavalamp

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:
```diff
@@ -428,9 +429,9 @@ func TestNoRestraint(t *testing.T) {
+		concurrencyLimit:       2,
 		evalDuration:           time.Second * 15,
-		expectedFair:           []bool{true},
+		expectedFair:           []bool{false},
```
To double check - this was simply a bug that it was marked as fair, right?

MikeSpreitzer replied:
Not exactly.  With the larger concurrency limit, the NoRestraint implementation yields the fair result.  The real bug was not testing with a small concurrency limit.

MikeSpreitzer replied:
"bug" is perhaps too strong a word here.  I think it is more interesting to test that the NoRestraint queueset is unfair when there is competition than it is to show that the fair result is delivered in the absence of competition.  However, there is some value in the latter --- it gives some degree of validation to test harness and its evaluation method.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:
```diff
@@ -746,38 +747,45 @@ func TestTooWide(t *testing.T) {
+	for _, testcase := range []struct {
+		margin2     float64
+		expectFair2 bool
+	}{{margin2: 0.26, expectFair2: true}, {margin2: 0.1, expectFair2: false}} {
```
nit: can we define the cases as we mostly try to do across codebase:
```
testCases := []struct {
	name        string
	margin2     float64
	expectFair2 bool
}{
  {name: "fair", margin2: 0.26, expectFair2: true},
  {name: "unfair", margin2: 0.1, expectFair2: false},
}

for _, testcase := range testCases {
```

MikeSpreitzer replied:
Are you concerned about having a name for the slice of test cases, or putting each case on its own line, or both?  I can imagine logic for the latter (making line-based edits independent), but the former is not obvious to me.

wojtek-t reviewed (commented):
Just some small nits.

MikeSpreitzer committed:
More test tweaks

Canonicalize listing of test cases.

Make TestNoRestraint try both cases: competition and none.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:
```diff
@@ -746,38 +759,48 @@ func TestTooWide(t *testing.T) {
+		expectFair2 bool
+	}{
+		{margin2: 0.26, expectFair2: true},
+		{margin2: 0.1, expectFair2: false}}
```
nit: for consistency with the other test:
```
   ...
   {margin2: 0.1, expectFair2: false},
}
```

Re name - I generally like having a name that already somewhat describes what the tests is supposed to test. But I can live with what you currently have too.

MikeSpreitzer replied:
The first point here is to use comma and linebreak after the last test case, right?

I am not sure I understand the second point.  Is it about the func name, TestWindup?  This func is about a scenario that has the windup problem.

MikeSpreitzer commented:
The force-push to 24cdf838ae2 adds the linebreak that I think was requested above and a comment explaining the `TestWindup` func.

MikeSpreitzer commented:
The force-push to 0ee1a7b4ff9 expands that comment a little.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:
```diff
@@ -744,40 +757,64 @@ func TestTooWide(t *testing.T) {
+		expectFair2 bool
+	}{
+		{margin2: 0.26, expectFair2: true},
+		{margin2: 0.1, expectFair2: false},
```
> I am not sure I understand the second point. Is it about the func name, TestWindup? This func is about a scenario that has the windup problem.

It's not about func name - it's about introducing an explicit "name" field in the testcase struct.
This allows a reader to quickly understand what a particular testcase is actually testing, which at least for me - even though I read those tests multiple times, I never remember what margin2 is for example.

wojtek-t commented:
/lgtm
/approve

caesarxuchao commented:
/triage accepted
